File: tutorial/titpack/01_lapack/sample-11.f
c>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
c        ACKNOWLEDGE THE USE OF THIS PACKAGE
c         WHEN YOU PUBLISH YOUR RESULTS !!!
c>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
c
c************ Sample main program #11 *****************
c  1d Heisenberg antiferromagnet with 8 spins
c    Eigenvalues and an eigenvector by diag
c*************************************************
      parameter (n=8,idim=70,ibond=n)
      implicit real*8 (a-h,o-z)
      dimension E(4)
      dimension elemnt(idim,idim)
      dimension list1(idim),list2(2,0:2**15)
      dimension bondwt(ibond),ipair(2*ibond),zrtio(ibond)
      dimension npair(2),sxx(1),szz(1)
      dimension wk(idim,8),iwk(idim),v(idim)
c
      data bondwt/ibond*-1.0d0/
      data zrtio/ibond*1.0d0/
      data ipair/1,2, 2,3, 3,4, 4,5, 5,6, 6,7, 7,8, 8,1/
c
      call sz(n,idim,0.0d0,list1,list2)
c- You may alternatively use szdy or sztn for faster processing -
c      call szdy(n,idim,0.0d0,list1,list2)
c  or
c      call sztn(n,idim,0.0d0,list1,list2)
c------------------------------------------------------------
      call elm3(n,idim,ipair,bondwt,zrtio,ibond,
     &          elemnt,idim,list1,list2)
c
      nvec=1
      ne=4
      call diag(elemnt,idim,idim,E,v,ne,nvec,wk,iwk)
      print 100,E
  100 format(/' [Eigenvalues]  '/2x,4f14.8)
c*** Do not forget to call elm3 again before calling check3
      call elm3(n,idim,ipair,bondwt,zrtio,ibond,
     &          elemnt,idim,list1,list2)
      call check3(elemnt,idim,idim,v,wk,Hexpec)
c
      npair(1)=1
      npair(2)=2
      call xcorr(n,idim,npair,1,v,sxx,list1,list2)
      print *,' sxx:',sxx
      call zcorr(n,idim,npair,1,v,szz,list1)
      print *,' szz:',szz
      end

File: tutorial/titpack/01_lapack_cxx/sample-11.cpp
/*****************************************************************************
*
* Rokko: Integrated Interface for libraries of eigenvalue decomposition
*
* Copyright (C) 2014 by Synge Todo <wistaria@comp-phys.org>
*
* Distributed under the Boost Software License, Version 1.0. (See accompanying
* file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
*
*****************************************************************************/

// C++ version of TITPACK Ver.2 by H. Nishimori

/************ Sample main program #11 *****************
* 1d Heisenberg antiferromagnet with 8 spins
* Eigenvalues and an eigenvector by diag
******************************************************/

#include "titpack.hpp"

int main() {
  std::cout.precision(10);

  // lattice structure
  int n = 8;
  int ibond = n;
  std::vector<int> ipair;
  for (int i = 0; i < ibond; ++i) {
    ipair.push_back(i);
    ipair.push_back((i + 1) % n);
  }

  // Hamiltonian parameters
  std::vector<double> bondwt(ibond, -1);
  std::vector<double> zrtio(ibond, 1);

  // table of configurations
  std::vector<int> list1;
  std::vector<std::vector<int> > list2;
  int idim = sz(n, 0, list1, list2);
  // You may alternatively use szdy or sztn for faster processing
  //   int = szdy(n, 0, list1, list2);
  // or
  //   int idim = sztn(n, 0, list1, list2);

  // Hamiltonian matrix
  matrix_type elemnt;
  elm3(n, ipair, bondwt, zrtio, elemnt, list1, list2);
  
  std::vector<double> E;
  matrix_type v;
  int nvec = 1;
  diag(elemnt, E, v, nvec);

  int ne = 4;
  std::cout << "[Eigenvalues]\n";
  for (int i = 0; i < ne; ++i) std::cout << '\t' << E[i];
  std::cout << std::endl;

  // // Do not forget to call elm3 again before calling check3
  elm3(n, ipair, bondwt, zrtio, elemnt, list1, list2);
  check3(elemnt, v, 0);
  
  std::vector<int> npair;
  npair.push_back(1);
  npair.push_back(2);
  std::vector<double> sxx(1);
  xcorr(n, npair, v, 0, sxx, list1, list2);
  std::cout << "sxx: " << sxx[0] << std::endl;
  std::vector<double> szz(1);
  zcorr(n, npair, v, 0, szz, list1);
  std::cout << "szz: " << szz[0] << std::endl;
}
